service/history/task/task.go

```diff
@@ -265,6 +265,12 @@ func (t *taskImpl) HandleErr(err error) (retErr error) {
 		return err
 	}
 
+	// If the shard were recently closed we just return an error, so we retry in a bit.
+	var errShardClosed *shard.ErrShardClosed
+	if errors.As(err, &errShardClosed) && time.Since(errShardClosed.ClosedAt) < shard.TimeBeforeShardClosedIsError {
```

Comment:
Rest of the changes look good in this iteration. I don't think we need this check here because once shard is closed it's closed. The queue/task components will not be re-used even if that same shard is re-assigned to this host

Reply:
This is the check that ensures that we do not emit error logs and metrics immediately when the shard is closed, causing all the noise. 
So we need to return early here, to avoid the noise.

Reply:
let's also consider making shard closed error non-retryable here https://github.com/uber/cadence/blob/b005d32d12c1c58a74acc46a40e84e567b7e7795/service/history/task/task.go#L323